## Import Libraries

In [1]:
#Importing libraries

import pandas as pd
import numpy as np
import os

## Import data 

In [2]:
#shortcut to data files
path = r'C:\Users\Sreelaya Narla\Desktop\PYTHON\2025 Instacart Basket Analysis'

In [3]:
#Import ords_prods_merge_extra_columns_4_7  pickle file from prepared data folder

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_extra_columns_4_7.pkl'))

In [4]:
#checking dimensions
ords_prods_merge.shape

(32404859, 18)

In [5]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  merge_status    price_range_loc     busiest_day busiest_days_summary  \
0         both  Mid-range product  Regularly busy         Regular days   
1         both  Mid-range product  Regularly busy         Regular days   
2         both  Low-range product  Regularly busy         Regular days   
3         both  Low-range product  Regularly busy         Regular days   
4         both  Low-range product  Regularly busy         Regular days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders

### Aggregated Mean Calculation for Entire Dataframe
### question 2
the aggregated mean of the “order_number” column grouped by “department_id”

In [6]:
ords_prods_merge.groupby('department_id').agg({ 'order_number' : ['mean'] })

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Analyze the result. How do the results for the entire dataframe differ from those of the subset?
### question 3
 Analysis of Mean Prices by Department

In general, the overall mean prices are slightly higher than those in the subset for most departments.

Department 1: Mean price increases from 14.80 (subset) to 15.46 (entire dataset).

Department 8: Drops from 16.46 (subset) to 15.34 in the full dataset — possibly indicating more lower-priced items in the full data.

Department 21 consistently has the highest average price in both cases, but again, it’s higher in the full dataset (22.90 vs 21.95).

Conclusion:
These differences suggest that the subset may not fully represent the overall pricing distribution in some departments. This emphasizes the importance of using the complete dataset when making business-wide pricing or strategy decisions.

## Create a loyalty flag 
### question 4

In [7]:
#step 1 -- Use transform() to create a new column that shows the maximum number of orders (order_number) each customer has made.

ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform('max')

In [8]:
#chcek output
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  merge_status    price_range_loc     busiest_day busiest_days_summary  \
0         both  Mid-range product  Regularly busy         Regular days   
1         both  Mid-range product  Regularly busy         Regular days   
2         both  Low-range product  Regularly busy         Regular days   
3         both  Low-range product  Regularly busy         Regular days   
4         both  Low-range product  Regularly busy         Regular days   

  busiest_period_of_day  max_order  
0        Average orders         10  
1        Average orders         10  
2        Average orders         10  
3        Average orders         10  
4        Average orders         10

In [9]:
#Step 2 -- .Use loc() to create another column that flags these customers as "loyal" based on a condition (e.g., more than 40 orders).
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [10]:
# Examining Value Counts on 'loyalty_flag'
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [11]:
#Checking the Updated Output
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                      NaN       14084                  2          0   
2                      NaN       12427                  3          0   
3                      NaN       26088                  4          0   
4                      NaN       26405                  5          0   
5                     15.0         196                  1          1   
6                     15.0       10258                  2          0   
7                     15.0       12427                  3          1   
8                     15.0       13176                  4          0   
9                     15.0       26088                  5          1   
10                    15.0       13032                  6          0   
11                    21.0         196                  1          1   
12                    21.0       12427                  2          1   
13                    21.0       10258                  3          1   
14                    21.0       25133                  4          0   

                               product_name  aisle_id  department_id  prices  \
0                                      Soda        77              7     9.0   
1   Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                       Original Beef Jerky        23             19     4.4   
3                Aged White Cheddar Popcorn        23             19     4.7   
4          XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                      Soda        77              7     9.0   
6                                Pistachios       117             19     3.0   
7                       Original Beef Jerky        23             19     4.4   
8                    Bag of Organic Bananas        24              4    10.3   
9                Aged White Cheddar Popcorn        23             19     4.7   
10                    Cinnamon Toast Crunch       121             14     4.0   
11                                     Soda        77              7     9.0   
12                      Original Beef Jerky        23             19     4.4   
13                               Pistachios       117             19     3.0   
14                    Organic String Cheese        21             16     8.6   

   merge_status    price_range_loc     busiest_day busiest_days_summary  \
0          both  Mid-range product  Regularly busy         Regular days   
1          both  Mid-range product  Regularly busy         Regular days   
2          both  Low-range product  Regularly busy         Regular days   
3          both  Low-

### difference between the spending habits of the three types of customers
### question 5

In [12]:
# Determining Basic Statistics of Spending Habits by Loyalty Category
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773487  1.0  25.0
New customer      7.801077  1.0  25.0
Regular customer  7.798145  1.0  25.0

##### All three groups have the same minimum and maximum prices, suggesting they have access to the same range of products.

The mean price is slightly lower for Loyal customers (7.77) compared to New (7.80) and Regular (7.80) customers.

This small difference could indicate that loyal customers are more price-conscious, possibly due to repeated purchases or familiarity with cheaper options.

While the difference is subtle, it does suggest that loyal customers may tend to spend slightly less per product on average than new or regular customers.

### Create a spending flag
### question 6

In [13]:
# Creating column for average spender with transform function
ords_prods_merge['average_spend']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\Sreelaya Narla\AppData\Local\Temp\ipykernel_19800\3981839492.py:2: FutureWarning: The provided callable <function mean at 0x0000027BB1A5F7E0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['average_spend']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [15]:
#Checking Relevant Columns 
ords_prods_merge[['user_id', 'average_spend']].head()

user_id  average_spend
0        1       6.367797
1        1       6.367797
2        1       6.367797
3        1       6.367797
4        1       6.367797

In [16]:
# Creating flag for low and high spender
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [17]:
# Check spender flags
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     32285150
High_spender      119709
Name: count, dtype: int64

In [18]:
pd.options.display.max_rows = 100

In [19]:
ords_prods_merge[['user_id', 'average_spend', 'spender_flag', 'prices']].head(100)

user_id  average_spend spender_flag  prices
0         1       6.367797  Low_spender     9.0
1         1       6.367797  Low_spender    12.5
2         1       6.367797  Low_spender     4.4
3         1       6.367797  Low_spender     4.7
4         1       6.367797  Low_spender     1.0
5         1       6.367797  Low_spender     9.0
6         1       6.367797  Low_spender     3.0
7         1       6.367797  Low_spender     4.4
8         1       6.367797  Low_spender    10.3
9         1       6.367797  Low_spender     4.7
10        1       6.367797  Low_spender     4.0
11        1       6.367797  Low_spender     9.0
12        1       6.367797  Low_spender     4.4
13        1       6.367797  Low_spender     3.0
14        1       6.367797  Low_spender     8.6
15        1       6.367797  Low_spender     3.2
16        1       6.367797  Low_spender     9.0
17        1       6.367797  Low_spender     4.4
18        1       6.367797  Low_spender     3.0
19        1       6.367797  Low_spender     8.6
20        1       6.367797  Low_spender     1.0
21        1       6.367797  Low_spender     9.0
22        1       6.367797  Low_spender     4.4
23        1       6.367797  Low_spender     3.0
24        1       6.367797  Low_spender     8.6
25        1       6.367797  Low_spender     2.7
26        1       6.367797  Low_spender    14.0
27        1       6.367797  Low_spender     5.0
28        1       6.367797  Low_spender    10.3
29        1       6.367797  Low_spender     9.0
30        1       6.367797  Low_spender     4.4
31        1       6.367797  Low_spender     3.0
32        1       6.367797  Low_spender     8.6
33        1       6.367797  Low_spender     9.0
34        1       6.367797  Low_spender     3.0
35        1       6.367797  Low_spender     4.4
36        1       6.367797  Low_spender     8.6
37        1       6.367797  Low_spender     4.0
38        1       6.367797  Low_spender     4.4
39        1       6.367797  Low_spender     9.0
40        1       6.367797  Low_spender     3.0
41        1       6.367797  Low_spender     8.6
42        1       6.367797  Low_spender    13.4
43        1       6.367797  Low_spender     1.8
44        1       6.367797  Low_spender     1.8
45        1       6.367797  Low_spender    13.4
46        1       6.367797  Low_spender     8.6
47        1       6.367797  Low_spender     9.0
48        1       6.367797  Low_spender     3.0
49        1       6.367797  Low_spender     4.4
50        1       6.367797  Low_spender     9.0
51        1       6.367797  Low_spender    13.4
52        1       6.367797  Low_spender     6.8
53        1       6.367797  Low_spender    12.6
54        1       6.367797  Low_spender     8.6
55        1       6.367797  Low_spender     3.0
56        1       6.367797  Low_spender     1.3
57        1       6.367797  Low_spender     4.0
58        1       6.367797  Low_spender     4.4
59        2       7.515897  Low_spender     5.2
60        2       7.515897  Low_spender     6.3
61        2       7.515897  Low_spender     2.3
62        2       7.515897  Low_spender    10.9
63        2       7.515897  Low_spender     2.7
64        2       7.515897  Low_spender     4.1
65        2       7.515897  Low_spender     5.0
66        2       7.515897  Low_spender     9.4
67        2       7.515897  Low_spender    14.8
68        2       7.515897  Low_spender     7.5
69        2       7.515897  Low_spender     4.4
70        2       7.515897  Low_spender    10.3
71        2       7.515897  Low_spender     2.5
72        2       7.515897  Low_spender     1.3
73        2       7.515897  Low_spender     2.0
74        2       7.515897  Low_spender     3.5
75        2       7.515897  Low_spender     6.3
76        2       7.515897  Low_spender     3.9
77        2       7.515897  Low_spender     2.0
78        2       7.515897  Low_spender     6.3
79        2       7.515897  Low_spender     5.2
80        2       7.515897  Low_spender     2.3
81        2       7.515897  Low_spender     6.6
82        2       7.

### Create an Order Frequency Flag
### question 7

In [20]:
#Create a column for frequent users
ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\Sreelaya Narla\AppData\Local\Temp\ipykernel_19800\2218061078.py:2: FutureWarning: The provided callable <function median at 0x0000027BB1B92A20> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [21]:
# Checking Relevant Columns 
ords_prods_merge[['user_id', 'days_since_prior_order', 'Customer_frequency']].head(15)

user_id  days_since_prior_order  Customer_frequency
0         1                     NaN                20.5
1         1                     NaN                20.5
2         1                     NaN                20.5
3         1                     NaN                20.5
4         1                     NaN                20.5
5         1                    15.0                20.5
6         1                    15.0                20.5
7         1                    15.0                20.5
8         1                    15.0                20.5
9         1                    15.0                20.5
10        1                    15.0                20.5
11        1                    21.0                20.5
12        1                    21.0                20.5
13        1                    21.0                20.5
14        1                    21.0                20.5

In [22]:
# Creating flag on order frequency
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] > 20, 'frequency_flag'] = 'Non_frequent_customer'

ords_prods_merge.loc[(ords_prods_merge['Customer_frequency']>10) & (ords_prods_merge['Customer_frequency']<=20),'frequency_flag'] = 'Regular customer'

ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [23]:
# Check frequency flags

ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent Customer        21559853
Regular customer          7208564
Non_frequent_customer     3636437
NaN                             5
Name: count, dtype: int64

In [24]:
ords_prods_merge[['user_id','days_since_prior_order', 'Customer_frequency', 'frequency_flag']].head(100)

user_id  days_since_prior_order  Customer_frequency         frequency_flag
0         1                     NaN                20.5  Non_frequent_customer
1         1                     NaN                20.5  Non_frequent_customer
2         1                     NaN                20.5  Non_frequent_customer
3         1                     NaN                20.5  Non_frequent_customer
4         1                     NaN                20.5  Non_frequent_customer
5         1                    15.0                20.5  Non_frequent_customer
6         1                    15.0                20.5  Non_frequent_customer
7         1                    15.0                20.5  Non_frequent_customer
8         1                    15.0                20.5  Non_frequent_customer
9         1                    15.0                20.5  Non_frequent_customer
10        1                    15.0                20.5  Non_frequent_customer
11        1                    21.0                20.5  Non_frequent_customer
12        1                    21.0                20.5  Non_frequent_customer
13        1                    21.0                20.5  Non_frequent_customer
14        1                    21.0                20.5  Non_frequent_customer
15        1                    21.0                20.5  Non_frequent_customer
16        1                    29.0                20.5  Non_frequent_customer
17        1                    29.0                20.5  Non_frequent_customer
18        1                    29.0                20.5  Non_frequent_customer
19        1                    29.0                20.5  Non_frequent_customer
20        1                    29.0                20.5  Non_frequent_customer
21        1                    28.0                20.5  Non_frequent_customer
22        1                    28.0                20.5  Non_frequent_customer
23        1                    28.0                20.5  Non_frequent_customer
24        1                    28.0                20.5  Non_frequent_customer
25        1                    28.0                20.5  Non_frequent_customer
26        1                    28.0                20.5  Non_frequent_customer
27        1                    28.0                20.5  Non_frequent_customer
28        1                    28.0                20.5  Non_frequent_customer
29        1                    19.0                20.5  Non_frequent_customer
30        1                    19.0                20.5  Non_frequent_customer
31        1                    19.0                20.5  Non_frequent_customer
32        1                    19.0                20.5  Non_frequent_customer
33        1                    20.0                20.5  Non_frequent_customer
34        1                    20.0                20.5  Non_frequent_customer
35        1                    20.0                20.5  Non_frequent_customer
36        1                    20.0                20.5  Non_frequent_customer
37        1                    20.0                20.5  Non_frequent_customer
38        1                    14.0                20.5  Non_frequent_customer
39        1                    14.0                20.5  Non_frequent_customer
40        1                    14.0                20.5  Non_frequent_customer
41        1                    14.0                20.5  Non_frequent_customer
42        1                    14.0                20.5  Non_frequent_customer
43        1                    14.0                20.5  Non_frequent_customer
44        1                     0.0                20.5  Non_frequent_customer
45        1                     0.0                20.5  Non_frequent_customer
46        1                     0.0                20.5  Non_frequent_customer
47        1                     0.0                20.5  Non_frequent_customer
48        1                     0.0                20.5  Non_frequent_customer
49        1                     0.0                20.

### Export the dataframe

In [25]:
#exporting the data frame
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'aggregated_ords_prods_merge.pkl'))